waku/README.md

```diff
@@ -0,0 +1,301 @@
+
+Waku nodes that do not support a particular packet code SHOULD ignore the packet without generating any error.
+
+Waku is a RLPx subprotocol called `waku` with version `0`. The version number corresponds to the major version in the header spec.
+
+### Packet Codes
+
+<!-- Packet code not clear -->
+
+The message codes reserved for Waku protocol: 0 - 127.
+
+<!-- Is this a MUST or rather a SHOULD -->
+
+Messages with unknown codes MUST be ignored, for forward compatibility of future versions.
+
+The Waku sub-protocol SHOULD support the following packet codes:
+
+| Name                       | Int Value |
+|----------------------------|-----------|
+| Status                     |     0     |
+| Messages                   |     1     |
+| PoW Requirement            |     2     |
+| Bloom Filter               |     3     |
+
+
+The following message codes are optional, but they are reserved for specific purpose.
+
+| Name                       | Int Value |
+|----------------------------|-----------|
+| P2P Request                |    126    |
+| P2P Message                |    127    |
+
+### Packet Format and Usage
+
+**Status** [`0`]
+
+<!-- @todo maybe make a table or something -->
+
+This packet contains two objects: integer message code (0x00) followed by a list of values: version, PoW requirement, and bloom filter, in this order. The bloom filter paramenter is optional; if it is missing or nil, the node is considered to be full node (i.e. accepts all messages). The format of PoW and bloom filter please see below (message codes 2 and 3).
+
+Status message should be sent after the initial handshake and prior to any other messages.
+
+**Messages** [`1`, `whisper_envelopes`]
+
+This packet contains two objects: message code (0x01) followed by a list (possibly empty) of Whisper Envelopes.
+
+This packet is used for sending the standard Whisper envelopes.
+
+**PoW Requirement** [`2`, `PoW`]
+
+This packet contains two objects: message code (0x02) followed by the value of PoW. This value is the IEEE 754 binary representation of 64-bit floating point number. Values of qNAN, sNAN, INF and -INF are not allowed. Negative values are also not allowed.
+
+This packet is used by Whisper nodes for dynamic adjustment of their individual PoW requirements. Recipient of this message should no longer deliver the sender messages with PoW lower than specified in this message.
+
+PoW is defined as average number of iterations, required to find the current BestBit (the number of leading zero bits in the hash), divided by message size and TTL:
+
+	PoW = (2**BestBit) / (size * TTL)
+
+PoW calculation:
+
+	fn short_rlp(envelope) = rlp of envelope, excluding env_nonce field.
+	fn pow_hash(envelope, env_nonce) = sha3(short_rlp(envelope) ++ env_nonce)
+	fn pow(pow_hash, size, ttl) = 2**leading_zeros(pow_hash) / (size * ttl)
+
+where size is the size of the full RLP-encoded envelope.
```

Comment:
We should probably change this:
```suggestion
where size is the size of the RLP-encoded envelope, excluding env_nonce field (size of `short_rlp(envelope)`).
```

As currently this is what is done in geth.
See also: https://github.com/ethereum/go-ethereum/pull/19753#pullrequestreview-253992108 

Using the full RLP-encoded envelope is more complex for the `bestBitTarget` calculation.


Note: This need to be updated also nim-eth, as there the size of the envelope object is used (also not in line with spec), see also https://github.com/status-im/nim-eth/blob/master/eth/p2p/rlpx_protocols/whisper/whisper_types.nim#L482